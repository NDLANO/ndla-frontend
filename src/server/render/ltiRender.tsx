/**
 * Copyright (c) 2024-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import config from "../../config";
import { getHtmlLang, getLocaleObject } from "../../i18n";
import { BAD_REQUEST, OK } from "../../statusCodes";
import { RenderFunc } from "../serverHelpers";

const bodyFields: Record<string, { required: boolean; value?: any }> = {
  lti_message_type: {
    required: true,
    value: ["basic-lti-launch-request", "ToolProxyRegistrationRequest", "ContentItemSelectionRequest"],
  },
  lti_version: { required: true, value: ["LTI-1p0", "LTI-2p0"] },
  launch_presentation_return_url: { required: false },
  launch_presentation_document_target: { required: false },
  launch_presentation_height: { required: false },
  launch_presentation_width: { required: false },
};

export function parseAndValidateParameters(body: any) {
  let validBody = true;
  const errorMessages: { field: string; message: string }[] = [];
  Object.keys(bodyFields).forEach((key) => {
    const bodyValue = body[key];
    if (bodyFields[key]?.required && !bodyValue) {
      validBody = false;
      errorMessages.push({ field: key, message: "Missing required field" });
      return;
    }
    if (bodyFields[key]?.value && !bodyFields[key]?.value.includes(bodyValue)) {
      errorMessages.push({
        field: key,
        message: `Value should be one of ${bodyFields[key]?.value}`,
      });
      validBody = false;
    }
  });
  return validBody
    ? {
        valid: true,
        ltiData: {
          ...body,
        },
      }
    : { valid: false, messages: errorMessages };
}

export const ltiRender: RenderFunc = async (req) => {
  const isPostRequest = req.method === "POST";
  const validParameters = isPostRequest ? parseAndValidateParameters(req.body) : undefined;
  if (isPostRequest) {
    if (!validParameters?.valid) {
      const messages = validParameters?.messages
        ?.map((msg) => `Field ${msg.field} with error: ${msg.message}.`)
        .join(",");
      return {
        status: BAD_REQUEST,
        data: { htmlContent: `Bad request. ${messages}` },
      };
    }
  }

  const lang = getHtmlLang(req.params.lang ?? "");
  const locale = getLocaleObject(lang).abbreviation;

  return {
    status: OK,
    data: {
      htmlContent: "",
      data: {
        initialProps: {
          ltiData: validParameters?.ltiData,
          locale,
        },
        config: config,
      },
    },
  };
};
