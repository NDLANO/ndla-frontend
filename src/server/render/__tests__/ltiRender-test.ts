/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { Request } from "express";
import { ltiRender, parseAndValidateParameters } from "../ltiRender";

test("ltiRender 200 OK ", async () => {
  const body = {
    lti_message_type: "basic-lti-launch-request",
    lti_version: "LTI-1p0",
    launch_presentation_return_url: "http://ndla-api/some-return-url",
    launch_presentation_document_target: "iframe",
    launch_presentation_height: "800",
    launch_presentation_width: "1200",
  };
  const response = await ltiRender(
    {
      params: {
        lang: "nb",
        articleId: "26050",
        resourceId: "urn:resource:123",
      },
      body,
      method: "POST",
      headers: {
        "user-agent": "Mozilla/5.0 Gecko/20100101 Firefox/58.0",
      },
    } as any as Request,
    {},
  );

  expect(response.status).toBe(200);
});

test("ltiRender 200 OK only required params", async () => {
  const body = {
    lti_message_type: "basic-lti-launch-request",
    lti_version: "LTI-1p0",
  };
  const response = await ltiRender(
    {
      params: {
        lang: "nb",
        articleId: "26050",
        resourceId: "urn:resource:123",
      },
      body,
      method: "POST",
      headers: {
        "user-agent": "Mozilla/5.0 Gecko/20100101 Firefox/58.0",
      },
    } as any as Request,
    {},
  );

  expect(response.status).toBe(200);
});

test("ltiRender 400 BAD REQUEST", async () => {
  const body = {
    lti_message_type: "basic-lti-launch-request",
    launch_presentation_return_url: "http://ndla-api/some-return-url",
    launch_presentation_document_target: "iframe",
    launch_presentation_height: "800",
    launch_presentation_width: "1200",
  };
  const response = await ltiRender(
    {
      params: {
        lang: "nb",
        articleId: "26050",
        resourceId: "urn:resource:123",
      },
      method: "POST",
      body,
      headers: {
        "user-agent": "Mozilla/5.0 Gecko/20100101 Firefox/58.0",
      },
    } as any as Request,
    {},
  );

  expect(response).toMatchSnapshot();
});

test("ltiRender 400 BAD REQUEST wrong values", async () => {
  const body = {
    lti_message_type: "basic-lti-launch-request",
    lti_version: "wrong version",
    launch_presentation_return_url: "http://ndla-api/some-return-url",
    launch_presentation_document_target: "iframe",
    launch_presentation_height: "800",
    launch_presentation_width: "1200",
  };
  const response = await ltiRender(
    {
      params: {
        lang: "nb",
        articleId: "26050",
        resourceId: "urn:resource:123",
      },
      method: "POST",
      body,
      headers: {
        "user-agent": "Mozilla/5.0 Gecko/20100101 Firefox/58.0",
      },
    } as any as Request,
    {},
  );

  expect(response).toMatchSnapshot();
});

test("parseAndValidateParameters no errors", () => {
  const body = {
    lti_message_type: "basic-lti-launch-request",
    lti_version: "LTI-1p0",
    launch_presentation_return_url: "http://ndla-api/some-return-url",
    launch_presentation_document_target: "iframe",
    launch_presentation_height: "800",
    launch_presentation_width: "1200",
  };
  const result = parseAndValidateParameters(body);
  expect(result).toMatchSnapshot();
});

test("parseAndValidateParameters errors", () => {
  const body = {};
  const result = parseAndValidateParameters(body);
  expect(result).toMatchSnapshot();
});
