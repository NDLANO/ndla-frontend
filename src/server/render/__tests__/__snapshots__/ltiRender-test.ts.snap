// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ltiRender 400 BAD REQUEST 1`] = `
{
  "data": {
    "htmlContent": "Bad request. Field lti_version with error: Missing required field.",
  },
  "status": 400,
}
`;

exports[`ltiRender 400 BAD REQUEST wrong values 1`] = `
{
  "data": {
    "htmlContent": "Bad request. Field lti_version with error: Value should be one of LTI-1p0,LTI-2p0.",
  },
  "status": 400,
}
`;

exports[`parseAndValidateParameters errors 1`] = `
{
  "messages": [
    {
      "field": "lti_message_type",
      "message": "Missing required field",
    },
    {
      "field": "lti_version",
      "message": "Missing required field",
    },
  ],
  "valid": false,
}
`;

exports[`parseAndValidateParameters no errors 1`] = `
{
  "ltiData": {
    "launch_presentation_document_target": "iframe",
    "launch_presentation_height": "800",
    "launch_presentation_return_url": "http://ndla-api/some-return-url",
    "launch_presentation_width": "1200",
    "lti_message_type": "basic-lti-launch-request",
    "lti_version": "LTI-1p0",
  },
  "valid": true,
}
`;
